File: src/main/java/ca/mcgill/science/tepid/server/rest/Destinations.java
package ca.mcgill.science.tepid.server.rest;

import ca.mcgill.science.tepid.common.Destination;
import ca.mcgill.science.tepid.common.Destination.DestinationTicket;
import ca.mcgill.science.tepid.common.Session;
import ca.mcgill.science.tepid.common.ViewResultSet;
import ca.mcgill.science.tepid.common.ViewResultSet.Row;
import ca.mcgill.science.tepid.server.util.CouchClient;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;

import javax.annotation.security.RolesAllowed;
import javax.ws.rs.*;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Path("/destinations")
public class Destinations {
    private final WebTarget couchdb = CouchClient.getTepidWebTarget();

    /**
     * Put destination map to CouchDb
     *
     * @param destinations map of destinations
     * @return post result
     */
    @PUT
    @RolesAllowed({"elder"})
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public String putDestinations(Map<String, Destination> destinations) {
        ObjectNode root = JsonNodeFactory.instance.objectNode();
        root.putArray("docs").addAll(new ObjectMapper().convertValue(destinations.values(), ArrayNode.class));
        return couchdb.path("_bulk_docs").request().post(Entity.entity(root, MediaType.APPLICATION_JSON)).readEntity(String.class);
    }

    /**
     * Retrieves map of room names as {@link String} and their details in {@link Destination}
     *
     * @param ctx context
     * @return Map of data
     */
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @RolesAllowed({"elder", "ctfer", "user"})
    public Map<String, Destination> getDestinations(@Context ContainerRequestContext ctx) {
        Session session = (Session) ctx.getProperty("session");
        List<Row<String, Destination>> rows = couchdb.path("_design/main/_view").path("destinations")
                .request(MediaType.APPLICATION_JSON).get(DestinationResultSet.class).rows;
        Map<String, Destination> out = new HashMap<>();
        for (Row<String, Destination> r : rows) {
            Destination d = r.value;
            if (!session.getRole().equals("elder")) {
                d.setDomain(null);
                d.setUsername(null);
                d.setPassword(null);
            }
            out.put(r.value.getId(), d);
        }
        return out;
    }

    @POST
    @Path("/{dest}")
    @RolesAllowed({"ctfer", "elder"})
    @Produces(MediaType.TEXT_PLAIN)
    @Consumes(MediaType.APPLICATION_JSON)
    public Response setStatus(@PathParam("dest") String id, DestinationTicket ticket, @Context ContainerRequestContext crc) {
        Session session = (Session) crc.getProperty("session");
        ticket.user = session.getUser();
        Destination dest = null;
        try {
            dest = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(Destination.class);
        } catch (Exception ignored) {
        }
        if (dest == null) return Response.status(404).entity("Could not find destination " + id).build();
        if (ticket.up) {
            dest.setUp(true);
            dest.setTicket(null);
        } else {
            dest.setUp(false);
            dest.setTicket(ticket);
        }
        couchdb.path(id).request().put(Entity.entity(dest, MediaType.APPLICATION_JSON));
        return Response.ok(id + " marked as " + (ticket.up ? "up" : "down")).build();
    }


    @DELETE
    @Path("/{dest}")
    @RolesAllowed("elder")
    @Produces(MediaType.APPLICATION_JSON)
    public String deleteQueue(@PathParam("dest") String destination) {
        String rev = couchdb.path(destination).request(MediaType.APPLICATION_JSON).get().readEntity(ObjectNode.class).get("_rev").asText();
        return couchdb.path(destination).queryParam("rev", rev).request().delete().readEntity(String.class);
    }

    private static class DestinationResultSet extends ViewResultSet<String, Destination> {
    }

}

File: src/main/java/ca/mcgill/science/tepid/server/rest/Destinations.kt
package ca.mcgill.science.tepid.server.rest

import ca.mcgill.science.tepid.models.bindings.CTFER
import ca.mcgill.science.tepid.models.bindings.ELDER
import ca.mcgill.science.tepid.models.bindings.USER
import ca.mcgill.science.tepid.models.data.Destination
import ca.mcgill.science.tepid.models.data.DestinationTicket
import ca.mcgill.science.tepid.models.data.FullDestination
import ca.mcgill.science.tepid.models.data.PutResponse
import ca.mcgill.science.tepid.server.db.DB
import ca.mcgill.science.tepid.server.db.remapExceptions
import ca.mcgill.science.tepid.server.util.failNotFound
import ca.mcgill.science.tepid.server.util.getSession
import ca.mcgill.science.tepid.server.util.logMessage
import org.apache.logging.log4j.kotlin.Logging
import javax.annotation.security.RolesAllowed
import javax.ws.rs.Consumes
import javax.ws.rs.DELETE
import javax.ws.rs.GET
import javax.ws.rs.POST
import javax.ws.rs.PUT
import javax.ws.rs.Path
import javax.ws.rs.PathParam
import javax.ws.rs.Produces
import javax.ws.rs.container.ContainerRequestContext
import javax.ws.rs.core.Context
import javax.ws.rs.core.MediaType

@Path("/destinations")
class Destinations {

    /**
     * Retrieves map of room names as [String] and their details in [Destination]
     *
     * @param ctx context
     * @return Map of data
     */
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @RolesAllowed(USER, CTFER, ELDER)
    fun getDestinations(@Context ctx: ContainerRequestContext): Map<String, Destination> {
        val session = ctx.getSession()
        return remapExceptions {
            DB.destinations.readAll()
                .mapNotNull {
                    val id = it._id ?: return@mapNotNull null
                    id to it.toDestination(session.role)
                }
                .toMap()
        }
    }

    @POST
    @RolesAllowed(ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun newDestination(destination: FullDestination): PutResponse =
        remapExceptions { DB.destinations.put(destination) }

    @GET
    @Path("/{dest}")
    @RolesAllowed(USER, CTFER, ELDER)
    fun getDestination(@Context ctx: ContainerRequestContext, @PathParam("dest") id: String): Destination {
        val session = ctx.getSession()
        return remapExceptions { DB.destinations.read(id).toDestination(session.role) }
    }

    @PUT
    @Path("/{dest}")
    @RolesAllowed(ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun putDestination(@PathParam("dest") id: String, destination: FullDestination): PutResponse =
        remapExceptions { DB.destinations.put(destination) }

    @DELETE
    @Path("/{dest}")
    @RolesAllowed(ELDER)
    fun deleteDestination(@PathParam("dest") id: String): Unit =
        remapExceptions { DB.destinations.deleteById(id) }

    @POST
    @Path("/{dest}/ticket")
    @RolesAllowed(CTFER, ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun setTicket(@PathParam("dest") id: String, ticket: DestinationTicket, @Context crc: ContainerRequestContext): String {
        val session = crc.getSession()
        ticket.user = session.user.toUser()
        val successText = "$id marked as ${if (ticket.up) "up" else "down"}"

        try {
            DB.destinations.update(id) {
                up = ticket.up
                this.ticket = if (ticket.up) null else ticket
                logger.info(
                    logMessage(
                        "destination status changed",
                        "status" to if (ticket.up) "up" else "down",
                        "reason" to ticket.reason,
                        "by" to ticket.user?.shortUser
                    )
                )
            }
        } catch (e: Exception) {
            failNotFound("Could not find destination $id")
        }
        return successText
    }

    @DELETE
    @Path("/{dest}/ticket")
    @RolesAllowed(CTFER, ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    fun deleteTicket(@PathParam("dest") destination: String): Unit =
        remapExceptions { DB.destinations.deleteById(destination) }

    private companion object : Logging
}
